RyuSocks.Test/Auth/UsernameAndPasswordTests.cs

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) RyuSOCKS
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+using RyuSocks.Auth;
+using RyuSocks.Packets.Auth;
+using RyuSocks.Packets.Auth.UsernameAndPassword;
+using RyuSocks.Test.Utils;
+using System;
+using System.Collections.Generic;
+using System.Security.Authentication;
+using Xunit;
+
+namespace RyuSocks.Test.Auth
+{
+    public class UsernameAndPasswordTests
+    {
+        [Theory]
+        [InlineData(new byte[] { 0xFF, 0xFF, 0xAA, 0x00, 0xCC })]
+        public void Authenticate_PackageWithWrongVersion(byte[] incomingPacket)
+        {
+            UsernameAndPassword usernameAndPassword = new()
+            {
+                Database = [],
+                IsClient = false
+            };
+            Assert.Throws<InvalidOperationException>(() => usernameAndPassword.Authenticate(incomingPacket, out _));
+        }
+
+        [Theory]
+        [UsernameAndPasswordRandomUsernameAndPasswords(1)]
```

Comment:
Since the review in #2 seems to be gone, I'll repeat what I said there:

I'm not really a fan of this. Ideally we want to test 3 different cases of input to a method:

- Well-formed valid data (for example an authentication request containing a valid username and password combination)
- Well-formed invalid data (for example an authentication request containing a valid username, but an invalid password)
- Malformed data (for example an empty packet or a packet full of `0xFF` bytes)

For every case you want to make sure that the result/outcome of your method is exactly what you expect it to be.

Note that you might need more than 1 test to cover one of the cases. For example the second case here needs multiple tests, since there are multiple ways a well-formed packet could be invalid.

Imo we should also avoid using `Random` in tests and use hardcoded inputs to test all of the cases I mentioned above.

Reply:
Malformed packages are caught in the first test (namely Authenticate_ThrowsOnWrongVersion).
Therefore there is little sense in testing for those in Authenticate_WorksOnRegisteredUserAuthentication.